### ISSUE TITLE
Expose `EmittedEvents` subtypes

### ISSUE THREAD

**Issue by farooqkz** [matrix-org/matrix-js-sdk#3506]
Currently it is a compound of these:

```
ClientEvent | RoomEvents | RoomStateEvents | CryptoEvents | MatrixEventEvents | RoomMemberEvents | UserEvents | CallEvent | Incoming | Incoming | Outgoing | Ended | Participants | SessionLoggedOut | NoConsent | BeaconEvent
```

But few of them are exported and many like `CallEvent`, `Incoming` and `RoomEvent` are not exported and therefore cannot be used.  And we have to hardcode their values in our apps rather than use these enums which is the right way.

**Comment by richvdh** [comment]
It's a general theme of matrix-js-sdk that it doesn't correctly export types that are meant to be public.

You can generally work around it by directly including the file where they are declared. For example, in the case of `CallEvent`, you can try:

```javascript
import type { CallEvent } from "matrix-js-sdk/src/webrtc/call";
```

**Comment by richvdh** [comment]
Totally agree. Importing things from individual files is just a workaround.

**Comment by davidisaaclee** [comment]
In my setup (Webpack 5 via create-react-app), importing an enum from a specific file (e.g. `import { ThreadEvent } from 'matrix-js-sdk/src/models/Thread'`) causes lots of build errors. I'm bad at JS build tools, so I'm not entirely sure what the problem is, but it sounds to me like the import is triggering Webpack to try to parse the referenced file, which fails.

I'm working around these issues by (1) using literal strings instead of enum members (`'Thread.new' as any`), and (2) maintaining a fork.

tbc, would PRs exposing some of these enums / types be welcome?

**Comment by richvdh** [comment]
Yes, I'd be very happy to review PRs exposing these enums/types.

**Comment by farooqkz** [comment]
> In my setup (Webpack 5 via create-react-app), importing an enum from a specific file (e.g. `import { ThreadEvent } from 'matrix-js-sdk/src/models/Thread'`) causes lots of build errors. I'm bad at JS build tools, so I'm not entirely sure what the problem is, but it sounds to me like the import is triggering Webpack to try to parse the referenced file, which fails.
> 
> I'm working around these issues by (1) using literal strings instead of enum members (`'Thread.new' as any`), and (2) maintaining a fork.
> 
> tbc, would PRs exposing some of these enums / types be welcome?

I have similar problems in my setup too when importing from individual SDK files. I'm using typescript to using strings directly is another problem...

**Comment by t3chguy** [comment]
@davidisaaclee you likely want to import it from `/lib/` rather than `/src/`

**Comment by richvdh** [comment]
> But few of them are exported and many like `CallEvent`, `Incoming` and `RoomEvent` are not exported and therefore cannot be used. And we have to hardcode their values in our apps rather than use these enums which is the right way.

Just as a matter of detail: `RoomEvent` is in fact exported correctly. But `CallEvent` and `Incoming` are indeed not.
